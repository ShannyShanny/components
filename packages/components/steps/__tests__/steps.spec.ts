import { mount } from '@vue/test-utils'
import { renderWork } from '@tests'
import IxStep from '../src/Step.vue'
import IxSteps from '../src/Steps.vue'
import { nextTick, Ref, ref } from 'vue'

const TestComponent = {
  components: { IxSteps, IxStep },
  template: `
  <IxSteps :active="1">
    <ix-step title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
    <ix-step title="In Progress" description="This is a description."></ix-step>
    <ix-step title="Waiting" description="This is a description."></ix-step>
  </IxSteps>
  `
}

describe('Steps.vue', () => {

  renderWork(TestComponent)

  test('active work', async () => {
    const active = ref(1) as Ref<number>
    const wrapper = mount({
      components: { IxSteps, IxStep },
      template: `
      <IxSteps :active="active">
        <ix-step title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
        <ix-step title="In Progress" description="This is a description."></ix-step>
        <ix-step title="Waiting" description="This is a description."></ix-step>
      </IxSteps>
      `,
      setup() {
        return { active }
      },
    })

    expect(wrapper.findAllComponents({ name: 'IxStep' })[active.value].classes()).toContain('ix-step-process')

    active.value = 2

    await nextTick()

    expect(wrapper.findAllComponents({ name: 'IxStep' })[active.value].classes()).toContain('ix-step-process')
  })  
  
  test('direction work', async () => {
    const active = ref(1) as Ref<number>
    const direction = ref('vertical') as Ref<string>
    const wrapper = mount({
      components: { IxSteps, IxStep },
      template: `
      <IxSteps :active="active" :direction="direction">
        <ix-step title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
        <ix-step title="In Progress" description="This is a description."></ix-step>
        <ix-step title="Waiting" description="This is a description."></ix-step>
      </IxSteps>
      `,
      setup() {
        return { active, direction }
      },
    })

    expect(wrapper.findAll('.ix-step-vertical').length).toBe(3)
  })

  test('percent work', async () => {
    const active = ref(1) as Ref<number>
    const percent = ref(25) as Ref<number>
    const wrapper = mount({
      components: { IxSteps, IxStep },
      template: `
      <IxSteps :active="active" :percent="percent">
        <ix-step title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
        <ix-step title="In Progress" description="This is a description."></ix-step>
        <ix-step title="Waiting" description="This is a description."></ix-step>
      </IxSteps>
      `,
      setup() {
        return { active, percent }
      },
    })

    expect(wrapper.findAll('.ix-step-head-percent-circle')[0].attributes().style).toBe('transform: rotate(135deg);');
  
    percent.value = 50
    await nextTick()
    expect(wrapper.findAll('.ix-step-head-percent-circle')[0].attributes().style).toBe('transform: rotate(225deg);');
  
  
    percent.value = 75
    await nextTick()
    expect(wrapper.findAll('.ix-step-head-percent-circle')[0].attributes().style).toBe('transform: rotate(225deg);');
    expect(wrapper.findAll('.ix-step-head-percent-circle')[1].attributes().style).toBe('transform: rotate(135deg);');
  
    percent.value = 100
    await nextTick()
    expect(wrapper.findAll('.ix-step-head-percent-circle')[0].attributes().style).toBe('transform: rotate(225deg);');
    expect(wrapper.findAll('.ix-step-head-percent-circle')[1].attributes().style).toBe('transform: rotate(225deg);');
  
  })

  test('direction work', async () => {
    const active = ref(1) as Ref<number>
    const progressDot = ref(true) as Ref<boolean>
    const wrapper = mount({
      components: { IxSteps, IxStep },
      template: `
      <IxSteps :active="active" :progressDot="progressDot">
        <ix-step title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
        <ix-step title="In Progress" description="This is a description."></ix-step>
        <ix-step title="Waiting" description="This is a description."></ix-step>
      </IxSteps>
      `,
      setup() {
        return { active, progressDot }
      },
    })

    expect(wrapper.findAll('.ix-step-dot').length).toBe(3)
  })
  
  test('size work', async () => {
    const active = ref(1) as Ref<number>
    const size = ref('small') as Ref<string>
    const wrapper = mount({
      components: { IxSteps, IxStep },
      template: `
      <IxSteps :active="active" :size="size">
        <ix-step title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
        <ix-step title="In Progress" description="This is a description."></ix-step>
        <ix-step title="Waiting" description="This is a description."></ix-step>
      </IxSteps>
      `,
      setup() {
        return { active, size }
      },
    })

    expect(wrapper.findAll('.ix-step-small').length).toBe(3)
  })

  test('steps(status) work', async () => {
    const active = ref(1) as Ref<number>
    const status = ref('error') as Ref<string>
    const wrapper = mount({
      components: { IxSteps, IxStep },
      template: `
      <IxSteps :active="active" :status="status">
        <ix-step title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
        <ix-step title="In Progress" description="This is a description."></ix-step>
        <ix-step title="Waiting" description="This is a description."></ix-step>
      </IxSteps>
      `,
      setup() {
        return { active, status }
      },
    })

    expect(wrapper.findAllComponents({ name: 'IxStep' })[active.value].classes()).toContain('ix-step-' + status.value)
  })

  test('icon work', async () => {
    const active = ref(1) as Ref<number>
    const wrapper = mount({
      components: { IxSteps, IxStep },
      template: `
      <IxSteps :active="active">
        <ix-step icon="user" title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
        <ix-step icon="verified" title="In Progress" description="This is a description."></ix-step>
        <ix-step icon="loading" title="Waiting" description="This is a description."></ix-step>
        <ix-step icon="smile" title="Waiting" description="This is a description."></ix-step>
      </IxSteps>
      `,
      setup() {
        return { active }
      },
    })

    expect(wrapper.find('.ix-icon-user').exists()).toBe(true)
    expect(wrapper.find('.ix-icon-verified').exists()).toBe(true)
    expect(wrapper.find('.ix-icon-loading').exists()).toBe(true)
    expect(wrapper.find('.ix-icon-smile').exists()).toBe(true)
  })

  test('step(status) work', async () => {
    const active = ref(1) as Ref<number>
    const status = ref('error') as Ref<string>
    const wrapper = mount({
      components: { IxSteps, IxStep },
      template: `
      <IxSteps :active="active">
        <ix-step title="Finish" sub-title="20:20:01" description="This is a description."></ix-step>
        <ix-step :status="status" title="In Progress" description="This is a description."></ix-step>
        <ix-step title="Waiting" description="This is a description."></ix-step>
      </IxSteps>
      `,
      setup() {
        return { active, status }
      },
    })

    
    expect(wrapper.findAllComponents({ name: 'IxStep' })[1].classes()).toContain('ix-step-' + status.value)
  })

})
