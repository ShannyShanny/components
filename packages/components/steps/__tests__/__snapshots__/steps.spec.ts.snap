// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps.vue render work 1`] = `"<div class=\\"ix-steps ix-steps--horizontal\\"><div class=\\"ix-step ix-step--medium ix-step--finish\\"><div class=\\"ix-step-tail\\"></div><div class=\\"ix-step-head\\"><!-- 进度环 --><!--v-if--><i class=\\"ix-icon ix-icon-check ix-step-head__icon\\" role=\\"img\\" arialabel=\\"check\\"></i></div><div class=\\"ix-step-content\\"><div class=\\"ix-step-title\\">Finish<span class=\\"ix-step-subtitle\\"><span>20:20:01</span></span></div><div class=\\"ix-step-description\\">This is a description.</div></div></div><div class=\\"ix-step ix-step--medium ix-step--process\\"><div class=\\"ix-step-tail\\"></div><div class=\\"ix-step-head\\"><!-- 进度环 --><!--v-if--><span class=\\"ix-step-head__text\\">2</span></div><div class=\\"ix-step-content\\"><div class=\\"ix-step-title\\">In Progress<span class=\\"ix-step-subtitle\\"><span></span></span></div><div class=\\"ix-step-description\\">This is a description.</div></div></div><div class=\\"ix-step ix-step--medium ix-step--wait\\"><div class=\\"ix-step-tail\\"></div><div class=\\"ix-step-head\\"><!-- 进度环 --><!--v-if--><span class=\\"ix-step-head__text\\">3</span></div><div class=\\"ix-step-content\\"><div class=\\"ix-step-title\\">Waiting<span class=\\"ix-step-subtitle\\"><span></span></span></div><div class=\\"ix-step-description\\">This is a description.</div></div></div></div>"`;
